src/providers/wfs/qgsbackgroundcachedfeatureiterator.cpp

```diff
@@ -255,25 +255,24 @@ void QgsThreadedFeatureDownloader::run()
 
 // -------------------------
 
-static QgsFeatureRequest addSubsetToFeatureRequest( const QgsFeatureRequest &requestIn,
-    const QgsBackgroundCachedSharedData *shared )
-{
-  if ( shared->clientSideFilterExpression().isEmpty() )
-  {
-    return requestIn;
-  }
-  QgsFeatureRequest requestOut( requestIn );
-  requestOut.combineFilterExpression( shared->clientSideFilterExpression() );
-  return requestOut;
-}
-
 QgsBackgroundCachedFeatureIterator::QgsBackgroundCachedFeatureIterator(
   QgsBackgroundCachedFeatureSource *source, bool ownSource,
   std::shared_ptr<QgsBackgroundCachedSharedData> shared,
   const QgsFeatureRequest &request )
-  : QgsAbstractFeatureIteratorFromSource<QgsBackgroundCachedFeatureSource>( source, ownSource, addSubsetToFeatureRequest( request, shared.get() ) )
+  : QgsAbstractFeatureIteratorFromSource<QgsBackgroundCachedFeatureSource>( source, ownSource, request )
   , mShared( shared )
 {
+  if ( !shared->clientSideFilterExpression().isEmpty() )
+  {
+    // backup current request because combine filter expression will remove the fid(s) filtering
+    if ( mRequest.filterType() == QgsFeatureRequest::FilterFid || mRequest.filterType() == QgsFeatureRequest::FilterFids )
+    {
+      mAdditionalRequest = mRequest;
+    }
+
+    mRequest.combineFilterExpression( shared->clientSideFilterExpression() );
```

Comment:
What happens if I do

```
request = QgsFeatureRequest();
request.setFilterExpression( 'area > 1000')
request.setFilterFid(1)
wfsLayer->getFeatures( request )
```

i.e. I think this needs to go into an else and the conditional part above modified accordingly

Reply:
In this case, the request is no longer an expression request, its a filter fid request (even if `mFilterExpression` is not reset, `mFilter` which is the request type is equals to FilterFid)

From what I understand, This is not possible to combine a filter fid and a filter expression in the same QgsFeatureRequest.

So, I would say that provider deals only with QgsFeatureRequest type, fid in your example.

Reply:
I think mRequest.combineFilterExpression will combine to `area > 1000 and [CONDITION CLIENT SIDE]` which might unexpectedly also filter fid 1.

Reply:
You 're right. I made the modifications